include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+        "destruct() must be unstoppable");
+    return schedule_destruct_(this, parent);
+  }
+
+  void register_child(_container_base* child) noexcept {
+    std::lock_guard lock{mutex_};
+    child->parent_ = this;
+    child->next_ = std::exchange(children_, child);
+  }
+
+  unifex::task<void> handle_construction_failure() noexcept {
+    switch (state()) {
+      case state_t::ALLOCATED: deleter_(this); break;
+      case state_t::DESTRUCT_SPAWNED: destructionEvt_.set(); break;
+      case state_t::CONSTRUCTED:
+        if (parent_ && parent_->state() == state_t::DESTROYING) {
+          co_await close_scope();
+          // bubble up cancellation, prevent returning an empty _ptr
+          co_await unifex::just_done();
+        }
+        break;
+      case state_t::DESTROYING:
+        // do nothing, ownership handed over to destroy()
+        break;
+    }
+  }
+};
+
+class async_scope_ref final {
+  unifex::v1::async_scope* scope_;
+  _container_base* container_;
+
+public:
+  /*implicit*/ async_scope_ref(unifex::v1::async_scope& scope) noexcept
+    : async_scope_ref(scope, nullptr) {}  // unmanaged scope
+
+  async_scope_ref(
+      unifex::v1::async_scope& scope, _container_base* container) noexcept
+    : scope_(&scope)
+    , container_(container) {}
+
+  void register_child(_container_base* child) noexcept {
+    UNIFEX_ASSERT(child != nullptr);
+    unifex::spawn_detached(child->destruct(container_), *scope_);
+  }
+
+  template <typename Sender>
+  [[nodiscard]] auto nest(Sender&& sender) noexcept(
+      noexcept(scope_->attach(static_cast<Sender&&>(sender))))
+      -> decltype(scope_->attach(static_cast<Sender&&>(sender))) {
+    return scope_->attach(static_cast<Sender&&>(sender));
+  }
+
+  friend bool operator==(async_scope_ref lhs, async_scope_ref rhs) noexcept {
+    return lhs.scope_ == rhs.scope_ && lhs.container_ == rhs.container_;
+  }
+
+  friend bool operator!=(async_scope_ref lhs, async_scope_ref rhs) noexcept {
+    return !(lhs == rhs);
+  }
+};
+
+template <typename Resource, typename Scheduler>
+struct _container final : _container_base {
+  _container(Scheduler scheduler) noexcept
+      : _container_base{destruct_this_resource,  //
+                        schedule_destruct,       //
+                        destroy_this_resource,   //
+                        deleter},
+        scheduler_(scheduler) {}
+
+  template <typename ResourceFactory>
+  auto construct(ResourceFactory&& factory) noexcept {
+    return unifex::on(
+        scheduler_,
+        unifex::just_from([factory = std::forward<ResourceFactory>(factory),
+                           this]() mutable noexcept {
+          // note for upstreaming: scope_ construction may succeed, but
+          // resource_ construction may fail
+          scope_.construct();
+          resource_.construct_with([&]() noexcept {
+            return factory(async_scope_ref{scope_.get(), this}, scheduler_);
+          });
+          return set_constructed();
```

Comment:
We're now upstreaming… We'll have to make `noexcept` conditional and handle exceptions.
```suggestion
                           this]() mutable noexcept(std::is_nothrow_invocable_v<ResourceFactory&, async_scope_ref, Scheduler&>) {
          scope_.construct();

          scope_guard cleanup = [&]() noexcept {
            scope_.destruct();

            // not actually sure what to do here... can we let
            // the exception bubble out?
          };

          resource_.construct_with([&]() {
            return factory(async_scope_ref{scope_.get(), this}, scheduler_);
          });
          
          cleanup.release();

          return set_constructed();
```